core/presence/presence.go

```diff
@@ -0,0 +1,343 @@
+		if value.Server != server {
+			entries = append(entries, value)
+		}
+	}
+
+	for _, value := range connections {
+		if value.Server != server {
+			return errors.Errorf("connection server mismatch, got %q expected %q", value.Server, server)
+		}
+		value.Status = Alive
+		value.LastSeen = r.clock.Now()
+		entries = append(entries, value)
+	}
+
+	r.entries = entries
+	return nil
+}
+
+// Connections returns all connections info that the recorder has.
+func (r *recorder) Connections() Connections {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	entries := make([]Value, len(r.entries))
+	copy(entries, r.entries)
+	return &connections{values: entries}
+}
+
+// ForModel returns connections just for the specified model
+func (c *connections) ForModel(model string) Connections {
+	var values []Value
+	for _, value := range c.values {
+		if value.Model == model {
+			values = append(values, value)
+		}
+	}
+	return &connections{model: model, values: values}
+}
+
+// ForServer returns connections that are connected to the specified server.
+func (c *connections) ForServer(server string) Connections {
+	var values []Value
+	for _, value := range c.values {
+		if value.Server == server {
+			values = append(values, value)
+		}
+	}
+	return &connections{model: c.model, values: values}
+}
+
+// ForAgent returns all the connections for the specified agent.
+func (c *connections) ForAgent(agent string) Connections {
+	var values []Value
+	for _, value := range c.values {
+		if value.Agent == agent {
+			values = append(values, value)
+		}
+	}
+	return &connections{model: c.model, values: values}
+}
+
+// Count returns the number of Values in the collection.
+func (c *connections) Count() int {
+	return len(c.values)
+}
+
+// Models returns all the models referred to in the connections.
+func (c *connections) Models() []string {
+	models := set.NewStrings()
+	for _, value := range c.values {
+		models.Add(value.Model)
+	}
+	return models.Values()
+}
+
+// Servers returns all the servers referred to in the connectinos.
+func (c *connections) Servers() []string {
+	servers := set.NewStrings()
+	for _, value := range c.values {
+		servers.Add(value.Server)
+	}
+	return servers.Values()
+}
+
+// Agents returns all the agents referred to in the connections.
+func (c *connections) Agents() []string {
+	agents := set.NewStrings()
+	for _, value := range c.values {
+		agents.Add(value.Agent)
+	}
+	return agents.Values()
+}
+
+// AgentStatus returns the status for the specified agent.
+func (c *connections) AgentStatus(agent string) (Status, error) {
+	if c.model == "" {
+		return Unknown, errors.New("connections not limited to a model, agent ambiguous")
+	}
+	result := Unknown
+	for _, value := range c.values {
+		if value.Agent == agent && !value.ControllerAgent {
+			if value.Status > result {
+				result = value.Status
+			}
+		}
+	}
+	return result, nil
+}
+
+// Values returns the Values for the connections.
+func (c *connections) Values() []Value {
+	return c.values
```

Comment:
we should be returning a copy lest the caller mutate the values

Reply:
Any change would only be changing the connections themselves not what is stored in the recorder, as a copy is already made for the Connections call.